# 🕵️‍♀️ 02. Análisis Exploratorio Preliminar: Clientes (Customer Details)

## 📋 ¿Por qué este análisis separado?
En nuestro proyecto tenemos dos fuentes de datos distintas. En el notebook anterior analizamos las **interacciones de las campañas de marketing** (`bank-additional.csv`). Ahora, vamos a explorar el segundo archivo: una base de datos demográfica de los **clientes** (`customer-details.xlsx`). 

El objetivo final de esta fase será cruzar ambos archivos (hacer un `MERGE`) para tener una visión 360º de a quién estamos llamando. Pero antes de unirlos, necesito auditar la calidad de este archivo Excel.

## 📂 Origen de los Datos y Metodología
* **Fuente:** Datos crudos en `../data/customer-details.xlsx`.
* **Herramienta Custom:** Volveré a utilizar mi módulo `src.sp_eda` para mantener la consistencia en la auditoría.

**¿Qué busco responder aquí?**
1. ¿Es este dataset lo suficientemente limpio como para cruzarlo directamente?
2. ¿Tenemos una clave primaria (`Primary Key`) única y válida para el cruce?
3. ¿Existen anomalías estadísticas en los ingresos o comportamiento web de los clientes?


In [1]:
# -------------------------------------------------------------------------
# PREPARACIÓN DEL ENTORNO Y LIBRERÍAS
# -------------------------------------------------------------------------
import pandas as pd
import sys

# Configuración de Pandas
pd.set_option('display.max_columns', None)

# Añado la raíz del proyecto al PATH para importar módulos propios
sys.path.append('../')

# Importo mi función de auditoría automatizada
from src import sp_eda as se

print("✅ Entorno configurado correctamente.")


✅ Entorno configurado correctamente.


In [2]:
# -------------------------------------------------------------------------
# CARGA DE DATOS Y EJECUCIÓN DEL EDA
# -------------------------------------------------------------------------
ruta_clientes = "../data/customer-details.xlsx"

try:
    # Nota: Uso read_excel porque el archivo de origen es .xlsx
    df_clientes = pd.read_excel(ruta_clientes)
    print("✅ Datos de clientes cargados con éxito. Iniciando auditoría...\n" + "="*50)
    
    # Ejecuto la auditoría
    se.eda_preliminar(df_clientes)
    
except FileNotFoundError:
    print(f"❌ Error: No se ha encontrado el archivo en la ruta {ruta_clientes}")


✅ Datos de clientes cargados con éxito. Iniciando auditoría...
--- 1. MUESTRA ALEATORIA ---


,Unnamed: 0,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
7998,7998,56740,0,1,2012-02-28,22,bdfbdb78-5e0d-4eef-88df-bc6d0f8ac2af
3637,3637,102407,0,2,2012-03-16,22,e1cfe496-d5b9-4971-bf39-1823766e9846
16960,16960,46139,1,2,2012-02-23,17,221a1118-db3e-43c8-a1d2-67e5f9a61adf
12907,12907,134984,1,0,2012-04-28,21,44063c9e-3c4e-46bb-910c-3881cf7d4700
2301,2301,163031,1,1,2012-04-25,27,c7c0062a-d669-4863-ab40-ee44b11a6dc9



--- 2. DIMENSIONES ---
Filas: 20115 | Columnas: 7

--- 3. INFO GENERAL ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20115 entries, 0 to 20114
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Unnamed: 0         20115 non-null  int64         
 1   Income             20115 non-null  int64         
 2   Kidhome            20115 non-null  int64         
 3   Teenhome           20115 non-null  int64         
 4   Dt_Customer        20115 non-null  datetime64[ns]
 5   NumWebVisitsMonth  20115 non-null  int64         
 6   ID                 20115 non-null  object        
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 1.1+ MB

--- 4. NULOS ---
✅ ¡Perfecto! No hay valores nulos.

--- 5. DUPLICADOS ---
✅ No hay filas duplicadas.

--- 6. ESTADÍSTICAS NUMÉRICAS ---


,count,mean,min,25%,50%,75%,max,std
Unnamed: 0,20115.0,10057.0,0.0,5028.5,10057.0,15085.5,20114.0,5806.844668
Income,20115.0,93087.213224,5852.0,49110.0,92962.0,136705.0,180791.0,50614.798255
Kidhome,20115.0,1.004176,0.0,0.0,1.0,2.0,2.0,0.816811
Teenhome,20115.0,0.999503,0.0,0.0,1.0,2.0,2.0,0.815298
Dt_Customer,20115,2012-07-01 17:15:14.362416128,2012-01-01 00:00:00,2012-04-01 00:00:00,2012-07-02 00:00:00,2012-09-30 00:00:00,2012-12-31 00:00:00,NaN
NumWebVisitsMonth,20115.0,16.54079,1.0,9.0,16.0,25.0,32.0,9.235324



--- 7. COLUMNAS DE TEXTO (Top 10 valores) ---

-> Columna: ID (Total únicos: 20115)


,Valor,Frecuencia
0,8299f0a6-e4c0-49fc-87ab-81308b70f80f,1
1,089b39d8-e4d0-461b-87d4-814d71e0e079,1
2,e9d37224-cb6f-4942-98d7-46672963d097,1
3,71100141-53d0-4913-9750-f730875080c7,1
4,afd96190-4abd-404c-87a6-74cc40caebec,1
5,1f04ebce-3ae8-4e7f-8306-30742a537173,1
6,094b2662-7c1f-47e6-86c8-ec406f5373fc,1
7,942c997e-9f20-41ff-98b4-d2ad4b728610,1
8,812d6c96-fef7-4c94-84b7-69d3cfd1eb4e,1
9,7ca2bbf3-e280-4493-a010-6a51de92bf9d,1


   (Se muestran solo los 10 primeros de 20115)


## 💡 Diagnóstico y Plan de Acción (Insights del EDA de Clientes)

¡Excelentes noticias! A diferencia del archivo del banco, este dataset está **extremadamente limpio**. Sin embargo, he tomado nota de varios puntos clave de cara a la preparación de los datos (ETL):

### 🟢 1. Calidad del Dato (Inmejorable)
* **Cero Nulos y Cero Duplicados:** Una rareza en el mundo real, pero que nos ahorrará mucho tiempo de imputación de datos. 
* **Fechas automáticas:** Al cargar un archivo `.xlsx`, Pandas ha sido capaz de inferir automáticamente que `Dt_Customer` es de tipo `datetime64[ns]`. No tendré que hacer el *parsing* manual que sí necesito para el otro archivo.

### 🔑 2. La Clave Primaria (ID)
* La columna `ID` tiene **20,115 valores únicos** para **20,115 filas**. 
* Esto es fundamental: confirma que `ID` es una clave primaria perfecta y sin colisiones. Será la columna sobre la cual haré el cruce (MERGE) más adelante para unir esta tabla con la del banco.

### 📊 3. Estadísticas y Formato
* **Nombres de columnas:** Tienen mayúsculas combinadas (`Income`, `Kidhome`). Para mantener el estándar *Clean Code*, las pasaré por mi función de limpieza para convertirlas a `snake_case`.
* **Índice residual:** La columna `Unnamed: 0` es basura generada al exportar el archivo original. La eliminaré directamente.
* **Coherencia de negocio:** Los datos tienen todo el sentido del mundo. Los ingresos van desde los 5K€ hasta los 180K€, y los hijos/adolescentes en casa (`Kidhome`, `Teenhome`) varían lógicamente entre 0 y 2.

**Próximo Paso 👉:** Ahora que tenemos las auditorías hechas, pasaremos a la fase de Limpieza (ETL). Para mantener el orden, lo dividiremos en dos fases: en el **03_limpieza_banco** limpiaremos el complejo dataset del Banco, y en el **04_limpieza_customer** haremos los ajustes ligeros de este dataset de Clientes y los uniremos.
